File: 211-design-add-and-search-words-data-structure/211-design-add-and-search-words-data-structure.js
class TrieNode {
    constructor(){
        this.children = {};
        this.end_of_word = false;   
    } 
}

var WordDictionary = function() {
    this.root = new TrieNode();
};

/** 
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function(word) {
    let curr = this.root;
    for (let ch of word){
        if (!(ch in curr.children)){
            curr.children[ch] = new TrieNode();
        }
        curr = curr.children[ch];
        
    }
    curr.end_of_word = true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function(word) {
    var dfs = function(idx, node) {
        let trie_node = node;
        for (let x = idx; x < word.length; x++) {
            let ch = word[x];
            if (ch === '.') {
                for (let child of Object.values(trie_node.children)) {
                    if (dfs(x+1, child) !== false) return true;
                }
                return false;
            } else {
                if (!(ch in trie_node.children)) return false;
                trie_node = trie_node.children[ch];
            }
        }
        return trie_node.end_of_word;
    }
    return dfs(0, this.root);
};

File: 211-design-add-and-search-words-data-structure/211-design-add-and-search-words-data-structure.cpp
struct Trie
{
    bool isWord;
    unordered_map<char, Trie*> Dictionary;
    Trie()
    {
        isWord = false;
    }
};

class WordDictionary
{
    Trie * dict;

public:
    WordDictionary()
    {
        dict = new Trie();
    }

    void addWord(string word)
    {
        Trie *itr = dict;
        for (auto c: word)
        {
            if (!itr->Dictionary[c])
                itr->Dictionary[c] = new Trie();
            itr = itr->Dictionary[c];
        }
        itr->isWord = true;
    }
    bool searchUtil(string word, Trie* &root, int index)
    {
        if(!root)
            return false;
        Trie *node = root;
        for (int i = index; i < word.length(); i++)
        {
            char c = word[i];
            if (c == '.')
            {
                for (auto itr = node->Dictionary.begin(); itr != node->Dictionary.end(); itr++)
                {
                    if (searchUtil(word, itr->second, i + 1))
                        return true;
                }
                return false;
            }
            else
            {
                if(!node->Dictionary[c])
                    return false;
                node = node->Dictionary[c];
            }
        }
        return node && node->isWord;
    }
    bool search(string word)
    {
        return searchUtil(word, dict, 0);
    }
};
